cmd/kubelet/app/server.go

```diff
@@ -376,46 +285,63 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
 		}
 	}
 
-	// Set feature gates based on the value in KubeletConfiguration
-	err = utilfeature.DefaultFeatureGate.Set(s.KubeletConfiguration.FeatureGates)
-	if err != nil {
-		return err
-	}
-
-	// Register current configuration with /configz endpoint
-	cfgz, cfgzErr := initConfigz(&s.KubeletConfiguration)
+	// Alpha Dynamic Configuration Implementation; this section only loads config from disk, it does not contact the API server
+	var ncc *nodeconfig.NodeConfigController
 	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) {
-		// Look for config on the API server. If it exists, replace s.KubeletConfiguration
-		// with it and continue. initKubeletConfigSync also starts the background thread that checks for new config.
-
-		// Don't do dynamic Kubelet configuration in runonce mode
-		if s.RunOnce == false {
-			remoteKC, err := initKubeletConfigSync(s)
-			if err == nil {
-				// Update s (KubeletServer) with new config from API server
-				s.KubeletConfiguration = *remoteKC
-				// Ensure that /configz is up to date with the new config
-				if cfgzErr != nil {
-					glog.Errorf("was unable to register configz before due to %s, will not be able to set now", cfgzErr)
-				} else {
-					setConfigz(cfgz, &s.KubeletConfiguration)
-				}
-				// Update feature gates from the new config
-				err = utilfeature.DefaultFeatureGate.Set(s.KubeletConfiguration.FeatureGates)
-				if err != nil {
-					return err
-				}
-			} else {
-				glog.Errorf("failed to init dynamic Kubelet configuration sync: %v", err)
-			}
+
+		// the NodeConfigDir must be specified to use the config controller
+		if len(s.NodeConfigDir) == 0 {
+			return fmt.Errorf("--node-config-dir must be specified to use the node config controller")
+		}
+		// compute absolute path for NodeConfigDir, if necessary
+		cfgDir := s.NodeConfigDir
+		if !filepath.IsAbs(cfgDir) {
+			cfgDir = filepath.Join(s.RootDirectory, s.NodeConfigDir)
+		}
+
+		// get the external representation of the bootstrap configuration
+		bootstrapKC := &componentconfigv1alpha1.KubeletConfiguration{}
+		err = api.Scheme.Convert(&s.KubeletConfiguration, bootstrapKC, nil)
+		if err != nil {
+			return err
+		}
+
+		// get the latest KubeletConfiguration checkpoint from disk, or load the init or default config if no checkpoints exist
+		var kcToUse *componentconfigv1alpha1.KubeletConfiguration
+		kcToUse, ncc, err = initNodeConfig(bootstrapKC, cfgDir)
+		if err != nil {
+			return err
+		}
+
+		// update s (KubeletServer) to any new config
```

Comment:
TODO here to hoist initial obtaining of config out of run() so the reassignment isn't required